src/Functions/splitByRegexp.cpp

```diff
@@ -148,11 +152,70 @@ class SplitByRegexpImpl
 
 using FunctionSplitByRegexp = FunctionTokens<SplitByRegexpImpl>;
 
+/// Fallback splitByRegexp to splitByChar when its 1st argument is a trivial char for better performance
+class SplitByRegexpOverloadResolver : public IFunctionOverloadResolver
+{
+public:
+    static constexpr auto name = "splitByRegexp";
+    static FunctionOverloadResolverPtr create(ContextPtr context) { return std::make_unique<SplitByRegexpOverloadResolver>(context); }
+
+    explicit SplitByRegexpOverloadResolver(ContextPtr context_) : context(context_), split_by_regexp(FunctionSplitByRegexp::create(context))
+    {
+    }
+
+    String getName() const override { return name; }
+    size_t getNumberOfArguments() const override { return SplitByRegexpImpl::getNumberOfArguments(); }
+    bool isVariadic() const override { return SplitByRegexpImpl::isVariadic(); }
+
+    FunctionBasePtr buildImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & return_type) const override
+    {
+        /// If the first argument is a trivial char, fallback splitByRegexp to splitByChar for better performance
+        if (couldFallbackToSplitByChar(arguments))
+            return FunctionFactory::instance().getImpl("splitByChar", context)->build(arguments);
+        else
+            return std::make_unique<FunctionToFunctionBaseAdaptor>(
+                split_by_regexp, collections::map<DataTypes>(arguments, [](const auto & elem) { return elem.type; }), return_type);
+    }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        return split_by_regexp->getReturnTypeImpl(arguments);
+    }
+
+private:
+    bool couldFallbackToSplitByChar(const ColumnsWithTypeAndName & arguments) const
+    {
+        const ColumnConst * col = checkAndGetColumnConstStringOrFixedString(arguments[0].column.get());
+        if (!col)
+            throw Exception(
+                ErrorCodes::ILLEGAL_COLUMN,
+                "Illegal column {} of first argument of function {}. "
+                "Must be constant string.",
+                arguments[0].column->getName(),
+                getName());
+
+        String pattern = col->getValue<String>();
+        if (pattern.size() == 1)
+        {
+            auto re = std::make_shared<OptimizedRegularExpression>(Regexps::createRegexp<false, false, false>(col->getValue<String>()));
+
+            std::string required_substring;
+            bool is_trivial;
+            bool required_substring_is_prefix;
+            re->getAnalyzeResult(required_substring, is_trivial, required_substring_is_prefix);
+            return is_trivial && required_substring == pattern;
```

Comment:
I tried, ut it will cause some ut failed. Sometimes is_trivial is true but `required_substring` != `pattern`. We should not fallback to splitByChar under such condition because it will generate a diff.